### ISSUE TITLE
Build dependency races

### ISSUE THREAD

**Issue by mblighsummit** [ninja-build/ninja#1460]
in this example, misc.o is complied from misc.c which includes thorn.h
I would expect this command to ALWAYS rebuild misc.o on the second call to ninja
ninja build/misc.o; touch include/thorn.h; ninja build/misc.o

If I run this quickly a few times in succession:

```
$ ninja build/misc.o; touch include/thorn.h; ninja build/misc.o
ninja: no work to do.
[1/1] gcc -Werror -g -rdynamic -std=gnu99 -D_GNU_SOURCE -DELQ_STANDALONE -O2 -Llib/ -Iinclude/ -fno-omit-frame-pointer -Wall -Wno-unused-result -Wno-unused-but-set-variable -fno-common -MMD -MF build/misc.o.d libthorn/misc.c -o build/misc.o -c

$ ninja build/misc.o; touch include/thorn.h; ninja build/misc.o
ninja: no work to do.
[1/1] gcc -Werror -g -rdynamic -std=gnu99 -D_GNU_SOURCE -DELQ_STANDALONE -O2 -Llib/ -Iinclude/ -fno-omit-frame-pointer -Wall -Wno-unused-result -Wno-unused-but-set-variable -fno-common -MMD -MF build/misc.o.d libthorn/misc.c -o build/misc.o -c

$ ninja build/misc.o; touch include/thorn.h; ninja build/misc.o
ninja: no work to do.
ninja: no work to do.
```

It seems to be timing-related, you have to do it fairly quickly.
Once it's in this state, no number of attempts will make it rebuild

```
$ ninja build/misc.o
ninja: no work to do.

$ ninja build/misc.o
ninja: no work to do.
```

running stat shows that the header file is indeed newer than the C file

```
$ stat build/misc.o include/thorn.h 
  File: 'build/misc.o'
  Size: 50776           Blocks: 104        IO Block: 4096   regular file
Device: 823h/2083d      Inode: 5036131     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  mbligh)   Gid: ( 1000/  mbligh)
Access: 2018-08-16 11:44:10.179483645 -0400
Modify: 2018-08-16 11:44:10.179483645 -0400
Change: 2018-08-16 11:44:10.179483645 -0400
 Birth: -
  File: 'include/thorn.h'
  Size: 17047           Blocks: 40         IO Block: 4096   regular file
Device: 823h/2083d      Inode: 5803040     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  mbligh)   Gid: ( 1000/  mbligh)
Access: 2018-08-16 11:44:10.703489603 -0400
Modify: 2018-08-16 11:44:10.703489603 -0400
Change: 2018-08-16 11:44:10.703489603 -0400
 Birth: -
```

My gut feeling was that it's somehow only looking at timestamps at a 1s granularity, but looking at your Stat() code, that doesn't seem to be obviously true:

```
#if defined(__APPLE__) && !defined(_POSIX_C_SOURCE)
  return ((int64_t)st.st_mtimespec.tv_sec * 1000000000LL +
          st.st_mtimespec.tv_nsec);
#elif (_POSIX_C_SOURCE >= 200809L || _XOPEN_SOURCE >= 700 || defined(_BSD_SOURCE) || defined(_SVID_SOURCE) || \
       defined(__BIONIC__) || (defined (__SVR4) && defined (__sun)))
  // For glibc, see "Timestamp files" in the Notes of http://www.kernel.org/doc/man-pages/online/pages/man2/stat.2.html
  // newlib, uClibc and musl follow the kernel (or Cygwin) headers and define the right macro values above.
  // For bsd, see https://github.com/freebsd/freebsd/blob/master/sys/sys/stat.h and similar
  // For bionic, C and POSIX API is always enabled.
  // For solaris, see https://docs.oracle.com/cd/E88353_01/html/E37841/stat-2.html.
  return (int64_t)st.st_mtim.tv_sec * 1000000000LL + st.st_mtim.tv_nsec;
#else
  return (int64_t)st.st_mtime * 1000000000LL + st.st_mtimensec;
#endif
```

Unless something is storing it in the ninja DB at second granularity? 
Might be something else entirely, but definitely not performing in the deterministic way I would expect.

**Comment by mathstuf** [comment]
Older Ninja releases did only look at timestamps with 1 second resolution. See #1219.

**Comment by QuLogic** [comment]
And note there are no _new_ releases that actually include that PR.

**Comment by jhasse** [comment]
As said, this was fixed by #1219. The only thing missing now is to get the release out ;)
